### PR TITLE
Pad

### PR DESCRIPTION
**Pull request by absolutelyNoWarranty**

### CODE REVIEW

File: strpipe/ops/pad.pyx [hunk 1]
```diff
@@ -12,43 +13,149 @@ from strpipe.data.types import STRING_LIST
...
+        self,
+        str sos_token=DefaultTokens.nul,
+        str eos_token=DefaultTokens.nul,
+        str pad_token=DefaultTokens.pad,
+        int maxlen=-1):
         self.input_type = STRING_LIST
         self.output_type = STRING_LIST
         self._pad_token = pad_token
+
+        if (sos_token == DefaultTokens.nul) != (eos_token == DefaultTokens.nul):
```
Comment (stegben) [inline strpipe/ops/pad.pyx]: Maybe wrap `token == DefaultTokens.nul` into a function `cstr_is_none`? I spent a lot of time to figureout what it does.

File: strpipe/ops/tests/test_pad.py [hunk 2]
```diff
@@ -1,15 +1,57 @@
...
 from strpipe.toolkit.default_tokens import DefaultTokens
 
 from strpipe.ops.pad import Pad
 
 
 def test_pad_correctly_created():
+    def yield_padders():
+        yield Pad()
+        yield Pad(pad_token='<pad>')
+        yield Pad(sos_token='<sos>', eos_token='<eos>')
```
Comment (stegben) [inline strpipe/ops/tests/test_pad.py]: add one more: `yield Pad(sos_token='<sos>', eos_token='<eos>', pad_token='<pad>')`

File: strpipe/ops/pad.pyx [hunk 3]
```diff
@@ -12,43 +13,149 @@ from strpipe.data.types import STRING_LIST
...
+        else:
+            raise ValueError("state must provide both start-of-sentence "
+                             "and end-of-sentence token not just one.")
+
+        if sos_token != DefaultTokens.nul:
+            input_data = input_data[: (maxlen - 2)]
+            sentences_with_boundary_tokens = add_start_end_token_in_sentences_in_c(
+                input_data,
+                sos_token,
+                eos_token)
```
[the remark this answers is not in the document]
Comment (absolutelyNoWarranty) [reply strpipe/ops/pad.pyx]: :ok_hand:

File: strpipe/ops/tests/test_pad.py (hunk 2 above)
Comment (absolutelyNoWarranty) [reply strpipe/ops/tests/test_pad.py]: :ok_hand:

File: strpipe/ops/pad.pyx (hunk 1 above)
Comment (absolutelyNoWarranty) [reply strpipe/ops/pad.pyx]: :ok_hand: 
Added `_check_tokens`

File: strpipe/ops/pad.pyx [hunk 4]
```diff
@@ -12,43 +13,174 @@ from strpipe.data.types import STRING_LIST
...
+    cdef str _eos_token
     cdef int _maxlen
 
-    def __init__(self, str pad_token=DefaultTokens.pad, int maxlen=-1):
+    def __init__(
+        self,
+        str sos_token=DefaultTokens.nul,
+        str eos_token=DefaultTokens.nul,
+        str pad_token=DefaultTokens.pad,
+        int maxlen=-1):
```
Comment (SoluMilken) [inline strpipe/ops/pad.pyx]: , newline

Comment (absolutelyNoWarranty) [reply strpipe/ops/pad.pyx]: :ok_hand: https://github.com/Yoctol/strpipe/pull/38/commits/4c1316491c8ccb41a41cbd626008f61b7128599f

**Comment by stegben** [comment]
bravo

**Pull request merged**